Hand out temporary loopback addresses from a larger pool

raboof opened the pull request:
Might have prevented https://github.com/akka/akka/issues/23528

raboof committed:
Hand out temporary loopback addresses from a larger pool

Might have prevented https://github.com/akka/akka/issues/23528

patriknw commented on akka-testkit/src/main/scala/akka/testkit/SocketUtil.scala:
```diff
@@ -25,21 +26,28 @@ object SocketUtil {
     temporaryServerAddresses(1, address, udp).head
 
-  def temporaryServerAddresses(numberOfAddresses: Int, hostname: String = "127.0.0.1", udp: Boolean = false): immutable.IndexedSeq[InetSocketAddress] = {
+  def temporaryServerAddresses(numberOfAddresses: Int, hostname: String = RANDOM_LOOPBACK_ADDRESS, udp: Boolean = false): immutable.IndexedSeq[InetSocketAddress] = {
```
interesting
search for "temporaryServerAddress", we sometimes specify the `hostname` and that should be removed

ktoso replied:
No strong opinion on that, good to merge as is I think

raboof commented on akka-testkit/src/main/scala/akka/testkit/SocketUtil.scala:
```diff
@@ -25,21 +26,28 @@ object SocketUtil {
     temporaryServerAddresses(1, address, udp).head
 
-  def temporaryServerAddresses(numberOfAddresses: Int, hostname: String = "127.0.0.1", udp: Boolean = false): immutable.IndexedSeq[InetSocketAddress] = {
+  def temporaryServerAddresses(numberOfAddresses: Int, hostname: String = RANDOM_LOOPBACK_ADDRESS, udp: Boolean = false): immutable.IndexedSeq[InetSocketAddress] = {
```
Good one - I noticed some tests disregard the host part and assume 'localhost', perhaps we should create a separate `temporaryLocalPort` for those use cases?

raboof commented:
Hmm I just noticed this code somewhere:
```
        try SocketUtil.temporaryServerAddress(address = "127.0.1.1", udp = proto == "udp")
        catch {
          case e: java.net.BindException ⇒
            info(s"skipping test due to [${e.getMessage}], you probably have to use `ifconfig lo0 alias 127.0.1.1`")
            pending
            null
        }
```

This suggests `127.0.1.1/8` isn't always a loopback address by default. Seems to work on my machine and Jenkins though

raboof committed:
Detect if binding on other loopback addresses works

Works for Linux, not for OSX

raboof committed:
Avoid hard-coding 'localhost' in more places

When the address is ignored and only a port is requested,
find a free port on 'localhost'

patriknw reviewed (approved):
LGTM
